src/priceClient/adapters/defiLlama.ts

```diff
@@ -0,0 +1,81 @@
+import assert from "assert";
+import { PriceFeedAdapter, TokenPrice } from "../priceClient";
+import { BaseHTTPAdapter } from "./baseAdapter";
+
+type DefiLlamaTokenPrice = {
+  price: number;
+  symbol: string;
+  timestamp: number;
+  confidence: number;
+};
+
+type DefiLlamaPriceResponse = {
+  coins: {
+    [key: string]: DefiLlamaTokenPrice;
+  };
+};
+
+const defaultTimeout = 5000; // mS
+
+export class PriceFeed extends BaseHTTPAdapter implements PriceFeedAdapter {
+  private _minConfidence: number;
+
+  constructor(name = "DefiLlama", { timeout, minConfidence = 0.9 }: { timeout?: number; minConfidence?: number }) {
+    super(name, "coins.llama.fi", { timeout: timeout ?? defaultTimeout });
+    this._minConfidence = minConfidence;
+  }
+
+  get minConfidence(): number {
+    return this._minConfidence;
+  }
+
+  set minConfidence(minConfidence: number) {
+    assert(minConfidence >= 0.0 && minConfidence <= 1.0);
+    this._minConfidence = minConfidence;
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    const price = await this.getPricesByAddress([address], currency);
+    return price[0];
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
+    if (currency != "usd") throw new Error(`Currency ${currency} not supported by DefiLlama`);
+
+    const path = "prices/current/" + addresses.map((address) => `ethereum:${address}`).join();
+    const tokenPrices: unknown = await this.query(path, {});
+    if (!this.validateResponse(tokenPrices))
+      throw new Error(`Unexpected ${this.name} response: ${JSON.stringify(tokenPrices)}`);
+
+    return addresses.map((addr: string) => {
+      const tokenPrice = tokenPrices.coins[`ethereum:${addr.toLowerCase()}`];
+      if (tokenPrice === undefined) throw new Error(`Token ${addr} missing from ${this.name} response`);
+
+      if (tokenPrice.confidence < this.minConfidence)
+        throw new Error(`Token ${addr} has low confidence score (${tokenPrice.confidence})`);
```

Comment:
Based on the description from their API docs, the confidence score seems to come into play when sourcing prices from fallback sources that _might_ be less reliable (i.e. directly from on-chain pools). This can effectively be used by consumers as a tripwire/shut-off valve in the event that DefiLlama continues to provide valid responses, even though the quality of its source data might be suspect.

I've set it to 0.9 per default in this implementation, but the caller can override the confidence score..so it's always an option to set it to 0.0 and effectively ignore the confidence score.

```
For tokens that we haven't been able to price in any other way, we find the pool with most liquidity for each on uniswap, curve and serum and then use the prices provided on those exchanges.

Unlike all the other tokens, we can't confirm that these prices are correct, so we only ingest the ones that have sufficient liquidity and, even in that case, we attach a confidence value to them that is related to the depth of liquidity and which represents our confidence in the quality of each price. API consumers can choose to filter out prices with low confidence values.
```